pre_commit/languages/download.py

```diff
@@ -0,0 +1,267 @@
+        return other.os == self.os and other.cpu == self.cpu
+
+
+class ChecksumError(ValueError):
+    def __init__(self, sri: SRI, message: str = 'Mismatched checksum'):
+        self.__sri = sri
+        self.__message = message
+
+    @property
+    def sri(self) -> SRI:
+        return self.__sri
+
+    @property
+    def message(self) -> str:
+        return self.__message
+
+    def __str__(self) -> str:
+        return f'''{self.__message}:
+ - expected: {self.sri.checksum}
+ - actual  : {self.sri.digest}'''
+
+
+class SRI:
+    def __init__(self, value: str):
+        self.__value = value
+        self._algorithm, self._checksum = self.value.split('-', 1)
+        if self.algorithm not in hashlib.algorithms_available:
+            raise ValueError('Invalid hashing algorithm.')
+
+        if b64encode(b64decode(self.checksum)).decode('utf-8') !=\
+                self.checksum:
+            raise ValueError('Invalid checksum string.')
+
+        self.__hasher = hashlib.new(self.algorithm)
+
+        if len(b64decode(self.checksum)) != self.__hasher.digest_size:
+            raise ValueError(
+                f'Invalid checksum string length for {self.algorithm}',
+            )
+
+    @property
+    def value(self) -> str:
+        return self.__value
+
+    @property
+    def algorithm(self) -> str:
+        return self._algorithm
+
+    @property
+    def checksum(self) -> str:
+        return self._checksum
+
+    @property
+    def digest(self) -> str:
+        return b64encode(self.__hasher.digest()).decode('utf-8')
+
+    def update(self, data: bytes) -> None:
+        return self.__hasher.update(data)
+
+    def check_content(self) -> None:
+        if self.digest != self.checksum:
+            raise ChecksumError(self)
+
+    def __str__(self) -> str:
+        return self.value
+
+
+@dataclass(frozen=True)
+class URI:
+    value: str
+
+    def __post_init__(self) -> None:
+        url = urlparse(self.value)
+        if not all([url.scheme, url.netloc]):
+            raise ValueError(f'Invalid URI: {self.value}')
+
+    def __str__(self) -> str:
+        return self.value
+
+
+@dataclass(frozen=True)
+class Metadata:
+    value: str
+
+    @property
+    def parts(self) -> tuple[str, str, str, str]:
+        first, second, third, fourth = self.value.splitlines()
+        return (first, second, third, fourth)
+
+    @property
+    def platform(self) -> Platform:
+        platform, _, _, _ = self.parts
+        return Platform(platform)
+
+    @property
+    def sri(self) -> SRI:
+        _, sri, _, _ = self.parts
+        return SRI(sri)
+
+    @property
+    def uri(self) -> URI:
+        _, _, uri, _ = self.parts
+        return URI(uri)
+
+    @property
+    def filename(self) -> PurePath:
+        _, _, _, path = self.parts
+        return PurePath(path)
+
+
+def download(uri: URI, sri: SRI, filename: Path) -> None:
+    filename.parent.mkdir(parents=True, exist_ok=True)
+    with urlopen(str(uri)) as ws:
```

Comment:
We need [`.netrc`](https://docs.python.org/3/library/netrc.html) parsing here.

_Something_ along the lines of:

```py
from netrc import netrc
from urllib.request import (
    HTTPBasicAuthHandler,
    HTTPPasswordMgrWithDefaultRealm,
    Request,
    build_opener,
)

def download(...):
    request = Request(str(uri))
    manager = HTTPPasswordMgrWithDefaultRealm()
    handler = HTTPBasicAuthHandler(manager)
    opener = build_opener(handler)
    for login, _, password in netrc().authenticators(str(uri)):
        manager.add_password(None, str(uri), login, password)
    with opener.open(request) as fp:
        ...
```

Reply:
I am a bit concerned with testing with netrc, because the HTTP response handler is custom rewritten right now to serve the content we want. It doesn't handle authentication.

Reply:
You don't really need to test any authentication, just that the `netrc` auth gets added to the request.

That is the HTTP request has:

> a request contains a header field in the form of Authorization: Basic <credentials>, where credentials is the [Base64](https://en.wikipedia.org/wiki/Base64) encoding of ID and password joined by a single colon `:`.